src/NServiceBus.Core/Persistence/InMemory/Gateway/ClientIdStorage.cs

```diff
@@ -0,0 +1,52 @@
+﻿namespace NServiceBus
+{
+    using System.Collections.Generic;
+
+    class ClientIdStorage
+    {
+        public ClientIdStorage(int maxSize)
+        {
+            this.maxSize = maxSize;
+        }
+
+        public bool IsDuplicate(string clientId)
+        {
+            if (clientIdSet.TryGetValue(clientId, out var existingNode)) // O(1)
+            {
+                clientIdList.Remove(existingNode); // O(1) operation, because we got the node reference
```

Comment:
I thought this would be safe but https://stackoverflow.com/questions/2043615/thread-safety-of-a-dictionarytkey-tvalue seems to suggest either adding locks of going with a concurrent dictionary?

Reply:
> A Dictionary can support multiple readers concurrently, as long as the collection is not modified.

there seems to be a good chance that we're modifying the collection concurrently while other threads may read it, so this should be protected with a lock for now (did not further think about concurrent dictionary, might also be a good solution)